types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,602 @@
+		wp_drm_lease_connector_v1_send_withdrawn(wl_resource);
+		wl_resource_set_user_data(wl_resource, NULL);
+		wl_list_remove(wl_resource_get_link(wl_resource));
+		wl_list_init(wl_resource_get_link(wl_resource));
+	}
+
+	wp_drm_lease_v1_send_lease_fd(lease->resource, fd);
+	close(fd);
+	return lease;
+}
+
+void wlr_drm_lease_request_v1_reject(
+		struct wlr_drm_lease_request_v1 *request) {
+	assert(request && request->lease);
+	wp_drm_lease_v1_send_finished(request->lease->resource);
+	request->invalid = true;
+}
+
+void wlr_drm_lease_v1_revoke(struct wlr_drm_lease_v1 *lease) {
+	assert(lease);
+	if (lease->resource != NULL) {
+		wp_drm_lease_v1_send_finished(lease->resource);
+	}
+
+	struct wlr_drm_lease_device_v1 *device = lease->device;
+	if (lease->lessee_id != 0) {
+		if (wlr_drm_backend_terminate_lease(device->backend,
+				lease->lessee_id) < 0) {
+			wlr_log_errno(WLR_DEBUG, "drm_terminate_lease");
+		}
+	}
+	struct wlr_drm_lease_connector_v1 *connector = lease->connector;
+	connector->active_lease = NULL;
+
+	struct wl_resource *resource;
+	wl_resource_for_each(resource, &device->resources) {
+		struct wl_client *client = wl_resource_get_client(resource);
+		drm_lease_connector_v1_send_to_client(connector, client, resource);
+	}
+}
+
+static void drm_lease_v1_destroy(struct wlr_drm_lease_v1 *lease) {
+	wlr_drm_lease_v1_revoke(lease);
+	free(lease);
+}
+
+static void drm_lease_v1_handle_resource_destroy(struct wl_resource *resource) {
+	struct wlr_drm_lease_v1 *lease = wlr_drm_lease_v1_from_resource(resource);
+	wl_list_remove(wl_resource_get_link(resource));
+	wl_list_init(wl_resource_get_link(resource));
+	lease->resource = NULL;
+	drm_lease_v1_destroy(lease);
+}
+
+static void drm_lease_v1_handle_destroy(
+		struct wl_client *client, struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static struct wp_drm_lease_v1_interface lease_impl = {
+	.destroy = drm_lease_v1_handle_destroy,
+};
+
+static void drm_lease_request_v1_destroy(struct wlr_drm_lease_request_v1 *req) {
+	if (!req) {
+		return;
+	}
+	free(req->connector);
+	free(req);
+}
+
+static void drm_lease_request_v1_handle_resource_destroy(
+		struct wl_resource *resource) {
+	struct wlr_drm_lease_request_v1 *req =
+		wlr_drm_lease_request_v1_from_resource(resource);
+	drm_lease_request_v1_destroy(req);
+	wl_list_remove(wl_resource_get_link(resource));
+	wl_list_init(wl_resource_get_link(resource));
+}
+
+static void drm_lease_request_v1_handle_request_connector(
+		struct wl_client *client, struct wl_resource *request_resource,
+		struct wl_resource *connector_resource) {
+	struct wlr_drm_lease_request_v1 *request =
+		wlr_drm_lease_request_v1_from_resource(request_resource);
+	struct wlr_drm_lease_connector_v1 *connector =
+		wlr_drm_lease_connector_v1_from_resource(connector_resource);
+
+	if (!connector) {
+		/* This connector offer has been withdrawn */
+		request->invalid = true;
+		return;
+	}
+
+	request->connector = connector;
```

Comment:
Discussed already on IRC, but to keep notes on it here as well: The protocol request `request_connector` is supposed to add the connector to a set of connectors, that the client *all together* later attempts to lease via a single `wp_drm_lease_v1` object.

Reply:
Took a while to figure out all the logic I dumbly removed thinking the protocol was upgraded to a single connector lease.